pkg/validate/strategies.go

```diff
@@ -27,48 +33,110 @@ type Strategy struct {
+			}
+			if err := s.validateBuildParams(buildStrategy.Spec.Parameters); err != nil {
 				return err
 			}
 		}
 	}
 	return nil
 }
 
-func (s Strategy) validateBuildStrategy(ctx context.Context, strategyName string, b *build.Build) error {
-	buildStrategy := &build.BuildStrategy{}
-	if err := s.Client.Get(ctx, types.NamespacedName{Name: strategyName, Namespace: b.Namespace}, buildStrategy); err != nil && !apierrors.IsNotFound(err) {
+func (s Strategy) validateBuildStrategy(ctx context.Context, strategyName string, buildStrategy *build.BuildStrategy) error {
+	if err := s.Client.Get(ctx, types.NamespacedName{Name: strategyName, Namespace: s.Build.Namespace}, buildStrategy); err != nil && !apierrors.IsNotFound(err) {
 		return err
 	} else if apierrors.IsNotFound(err) {
-		b.Status.Reason = build.BuildStrategyNotFound
-		b.Status.Message = fmt.Sprintf("buildStrategy %s does not exist in namespace %s", b.Spec.Strategy.Name, b.Namespace)
+		s.Build.Status.Reason = build.BuildStrategyNotFound
+		s.Build.Status.Message = fmt.Sprintf("buildStrategy %s does not exist in namespace %s", s.Build.Spec.Strategy.Name, s.Build.Namespace)
 	}
 
 	return nil
 }
 
-func (s Strategy) validateClusterBuildStrategy(ctx context.Context, strategyName string, b *build.Build) error {
-	clusterBuildStrategy := &build.ClusterBuildStrategy{}
+func (s Strategy) validateClusterBuildStrategy(ctx context.Context, strategyName string, clusterBuildStrategy *build.ClusterBuildStrategy) error {
 	if err := s.Client.Get(ctx, types.NamespacedName{Name: strategyName}, clusterBuildStrategy); err != nil && !apierrors.IsNotFound(err) {
 		return err
 	} else if apierrors.IsNotFound(err) {
-		b.Status.Reason = build.ClusterBuildStrategyNotFound
-		b.Status.Message = fmt.Sprintf("clusterBuildStrategy %s does not exist", b.Spec.Strategy.Name)
+		s.Build.Status.Reason = build.ClusterBuildStrategyNotFound
+		s.Build.Status.Message = fmt.Sprintf("clusterBuildStrategy %s does not exist", s.Build.Spec.Strategy.Name)
 	}
 	return nil
 }
+
+func (s Strategy) validateBuildParams(parameters []build.Parameter) error {
+
+	if len(s.Build.Spec.Params) == 0 {
+		return nil
+	}
+
+	// Check that the Build param is defined in the strategies parameters
+	s.validateParamsInStrategies(s.Build.Spec.Params, parameters)
+
+	// Check that the Build param is not a restriced shipwright one
+	s.validateParamsNamesDefinition()
+
+	return nil
+}
+
+func (s Strategy) validateParamsNamesDefinition() {
+
+	undesiredParams := []string{}
+
+	for _, p := range s.Build.Spec.Params {
+		if m := systemReservedParametersRegexp.MatchString(p.Name); m {
+			undesiredParams = append(undesiredParams, p.Name)
+		}
+	}
+
+	if len(undesiredParams) > 0 {
+		s.Build.Status.Reason = build.RestrictedParametersInUse
+		s.Build.Status.Message = fmt.Sprintf("restricted parameters in use: %s", strings.Join(undesiredParams, ","))
+	}
+}
+
+func (s Strategy) validateParamsInStrategies(params []build.Param, parameters []build.Parameter) {
```

Comment:
I'm not a huge fan of there being both a `Param` and a `Parameter` type (I'm also not wild about Tekton's `Param` and `ParamSpec` types 😒)

That being said, based on my understanding of our backward-compatibility requirements, we can change these Go type names when/if we come up with better names, so long as the YAML representation remains compatible.

Reply:
alright, lets do that later if needed.